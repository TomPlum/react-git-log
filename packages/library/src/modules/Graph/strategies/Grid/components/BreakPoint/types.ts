import { CSSProperties } from 'react'
import { BreakPointTheme } from 'hooks/useTheme'

export interface BreakPointProps {
  position: BreakPointPosition
  className?: string
  color: string
  style?: Partial<Record<BreakPointTheme, CSSProperties>>
}

export type BreakPointPosition = 'top' | 'bottom'