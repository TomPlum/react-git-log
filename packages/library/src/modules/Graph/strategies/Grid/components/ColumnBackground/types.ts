export interface ColumnBackgroundProps {
  id: string
  index: number
  colour: string
  commitNodeIndex: number
}