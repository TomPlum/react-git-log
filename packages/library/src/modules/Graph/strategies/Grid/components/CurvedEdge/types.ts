export interface CurvedEdgeProps {
  id: string
  colour: string
  path: string
  dashed?: boolean
}