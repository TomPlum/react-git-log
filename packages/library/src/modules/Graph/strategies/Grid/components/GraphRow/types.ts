import { Commit } from 'types/Commit'
import { GraphColumnState } from 'modules/Graph/strategies/Grid/components/GraphColumn'

export interface GraphRowProps {
  id: number
  commit: Commit
  columns: GraphColumnState[]
}