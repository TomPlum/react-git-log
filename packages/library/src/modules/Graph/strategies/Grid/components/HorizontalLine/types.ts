import { GraphColumnState } from 'modules/Graph/strategies/Grid/components/GraphColumn'

export interface HorizontalLineProps {
  columnIndex: number
  commitNodeIndex: number
  columnColour: string
  state: GraphColumnState
}