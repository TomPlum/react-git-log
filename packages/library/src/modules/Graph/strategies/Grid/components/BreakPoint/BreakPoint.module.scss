.Slash {
  position: absolute;
  width: 14px;
  height: 3px;
  border-radius: 2px;
  background: var(--breakpoint-colour);
  transform: translate(-50%, -50%) rotate(-45deg);
  transform-origin: center;
  left: calc(50% + 2px);

  &--top {
    top: 0;
  }

  &--bottom {
    top: 100%;
  }
}

.Dot {
  position: absolute;
  $diameter: 8px;
  width: $diameter;
  height: $diameter;
  border-radius: 50%;
  background: var(--breakpoint-colour);
  transform: translate(-50%, -50%);
  left: calc(50% + 2px);

  &--top {
    top: 0;
  }

  &--bottom {
    top: 100%;
  }
}

.Ring {
  position: absolute;
  $diameter: 6px;
  width: $diameter;
  height: $diameter;
  border-radius: 50%;
  border: 2px solid var(--breakpoint-colour);
  transform: translate(-50%, -50%);
  transform-origin: center;
  left: calc(50% + 2px);

  &--top {
    top: -4px;
  }

  &--bottom {
    top: calc(100% + 4px);
  }
}

.ZigZag {
  position: absolute;

  // The first diagonal line closest to the node
  &::before {
    position: absolute;
    content: '';
    width: 5px;
    height: 2px;
    background: var(--breakpoint-colour);
    transform: translate(-50%, -50%) rotate(-50deg);
    transform-origin: center;
  }

  // The second diagonal line farthest away from the node
  &::after {
    position: absolute;
    content: '';
    width: 5px;
    height: 2px;
    background: var(--breakpoint-colour);
    transform: translate(-50%, -50%) rotate(50deg);
    transform-origin: center;
  }

  &--top {
    top: 0;

    // The first diagonal line closest to the node
    &::before {
      top: -1px;
      left: calc(50% + 4px);
    }

    // The second diagonal line farthest away from the node
    &::after {
      top: -5px;
      left: calc(50% + 5px);
    }
  }

  &--bottom {
    top: 100%;

    // The first diagonal line closest to the node
    &::before {
      top: 2px;
      left: calc(50% + 2px);
    }

    // The second diagonal line farthest away from the node
    &::after {
      top: 6px;
      left: 50%;
    }
  }
}

.Line {
  position: absolute;
  width: 12px;
  height: 2px;
  border-radius: 10px;
  background: var(--breakpoint-colour);
  left: calc(50% - 4px);

  &--top {
    top: 0;
  }

  &--bottom {
    top: 100%;
  }
}

.DoubleLine {
  position: absolute;

  &::before {
    position: absolute;
    content: '';
    width: 14px;
    height: 2px;
    border-radius: 10px;
    background: var(--breakpoint-colour);
    left: calc(50% - 4px);
  }

  &::after {
    position: absolute;
    content: '';
    width: 8px;
    height: 1px;
    border-radius: 10px;
    background: var(--breakpoint-colour);
    left: calc(50% - 1px);
  }

  &--top {
    top: 0;

    &::after {
      top: -4px;
    }
  }

  &--bottom {
    top: 100%;

    &::after {
      top: 4px;
    }
  }
}