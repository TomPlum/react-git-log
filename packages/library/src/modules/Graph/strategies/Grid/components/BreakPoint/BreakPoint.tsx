import { BreakPointProps } from './types'
import classNames from 'classnames'
import classes from './BreakPoint.module.scss'
import { useGraphContext } from 'modules/Graph/context'
import { CSSProperties, useMemo } from 'react'

export const BreakPoint = ({ position, className, color, style }: BreakPointProps) => {
  const { breakPointTheme } = useGraphContext()

  const commonStyles = useMemo<CSSProperties>(() => ({
    '--breakpoint-colour': color
  } as CSSProperties), [color])

  switch (breakPointTheme) {
    case 'slash': {
      return (
        <div
          style={{ ...commonStyles, ...style?.slash }}
          data-testid={`graph-break-point-slash-${position}`}
          className={classNames(
            classes.Slash,
            classes[`Slash--${position}`],
            className
          )}
        />
      )
    }
    case 'dot': {
      return (
        <div
          style={{ ...commonStyles, ...style?.dot }}
          data-testid={`graph-break-point-dot-${position}`}
          className={classNames(
            classes.Dot,
            classes[`Dot--${position}`],
            className
          )}
        />
      )
    }
    case 'ring': {
      return (
        <div
          style={{ ...commonStyles, ...style?.ring }}
          data-testid={`graph-break-point-ring-${position}`}
          className={classNames(
            classes.Ring,
            classes[`Ring--${position}`],
            className
          )}
        />
      )
    }
    case 'zig-zag': {
      return (
        <div
          style={{ ...commonStyles, ...style?.['zig-zag'] }}
          data-testid={`graph-break-point-zig-zag-${position}`}
          className={classNames(
            classes.ZigZag,
            classes[`ZigZag--${position}`],
            className
          )}
        />
      )
    }
    case 'line': {
      return (
        <div
          style={{ ...commonStyles, ...style?.['line'] }}
          data-testid={`graph-break-point-line-${position}`}
          className={classNames(
            classes.Line,
            classes[`Line--${position}`],
            className
          )}
        />
      )
    }
    case 'double-line': {
      return (
        <div
          style={{ ...commonStyles, ...style?.['double-line'] }}
          data-testid={`graph-break-point-double-line-${position}`}
          className={classNames(
            classes.DoubleLine,
            classes[`DoubleLine--${position}`],
            className
          )}
        />
      )
    }
    case 'arrow': {
      return (
        <div
          style={{ ...commonStyles, ...style?.arrow }}
          data-testid={`graph-break-point-arrow-${position}`}
          className={classNames(
            classes.Arrow,
            classes[`Arrow--${position}`],
            className
          )}
        />
      )
    }
  }
}