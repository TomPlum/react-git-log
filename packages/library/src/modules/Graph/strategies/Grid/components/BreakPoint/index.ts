export * from './types'
export * from './BreakPoint'