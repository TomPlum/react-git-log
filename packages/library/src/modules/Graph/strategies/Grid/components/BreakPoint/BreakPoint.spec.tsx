import { render, screen } from '@testing-library/react'
import { BreakPoint } from './BreakPoint'
import * as graphContext from 'modules/Graph/context'
import { graphContextBag } from 'test/stubs'
import { BreakPointTheme } from 'hooks/useTheme'
import { BreakPointPosition } from './types'

describe('Filtered Graph Break Point', () => {
  const variations: { theme: BreakPointTheme, position: BreakPointPosition }[] = [
    {
      theme: 'dot',
      position: 'top'
    },
    {
      theme: 'dot',
      position: 'bottom'
    },
    {
      theme: 'slash',
      position: 'top'
    },
    {
      theme: 'slash',
      position: 'bottom'
    },
    {
      theme: 'zig-zag',
      position: 'top'
    },
    {
      theme: 'zig-zag',
      position: 'bottom'
    },
    {
      theme: 'line',
      position: 'top'
    },
    {
      theme: 'line',
      position: 'bottom'
    },
    {
      theme: 'double-line',
      position: 'top'
    },
    {
      theme: 'double-line',
      position: 'bottom'
    },
    {
      theme: 'ring',
      position: 'top'
    },
    {
      theme: 'ring',
      position: 'bottom'
    },
    {
      theme: 'arrow',
      position: 'top'
    },
    {
      theme: 'arrow',
      position: 'bottom'
    }
  ]

  variations.forEach(({ position, theme }) => {
    it(`should render a [${position}] break point correctly when the theme is [${theme}]`, () => {
      vi.spyOn(graphContext, 'useGraphContext').mockReturnValue(graphContextBag({
        breakPointTheme: theme
      }))

      const { asFragment } = render(
        <BreakPoint
          position={position}
          color='rbg(125, 67, 100)'
        />
      )

      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('Style Overriding', () => {
    variations.forEach(({ position, theme }) => {
      it('should spread styles props into the underlying element wrapper for the matching theme', () => {
        vi.spyOn(graphContext, 'useGraphContext').mockReturnValue(graphContextBag({
          breakPointTheme: theme
        }))

        render(
          <BreakPoint
            color='green'
            position={position}
            style={{
              [theme]: {
                borderColor: 'antiquewhite'
              }
            }}
          />
        )

        expect(screen.getByTestId(`graph-break-point-${theme}-${position}`)).toHaveStyle({
          borderColor: 'antiquewhite'
        })
      })
    })
  })
})