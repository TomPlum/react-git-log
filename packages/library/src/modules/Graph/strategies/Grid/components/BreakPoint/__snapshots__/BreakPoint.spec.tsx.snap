// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Filtered Graph Break Point > should render a [bottom] break point correctly when the theme is [dot] 1`] = `
<DocumentFragment>
  <div
    class="Dot Dot--bottom"
    data-testid="graph-break-point-dot-bottom"
    style="--breakpoint-colour: rbg(125, 67, 100);"
  />
</DocumentFragment>
`;

exports[`Filtered Graph Break Point > should render a [bottom] break point correctly when the theme is [double-line] 1`] = `
<DocumentFragment>
  <div
    class="DoubleLine DoubleLine--bottom"
    data-testid="graph-break-point-double-line-bottom"
    style="--breakpoint-colour: rbg(125, 67, 100);"
  />
</DocumentFragment>
`;

exports[`Filtered Graph Break Point > should render a [bottom] break point correctly when the theme is [line] 1`] = `
<DocumentFragment>
  <div
    class="Line Line--bottom"
    data-testid="graph-break-point-line-bottom"
    style="--breakpoint-colour: rbg(125, 67, 100);"
  />
</DocumentFragment>
`;

exports[`Filtered Graph Break Point > should render a [bottom] break point correctly when the theme is [ring] 1`] = `
<DocumentFragment>
  <div
    class="Ring Ring--bottom"
    data-testid="graph-break-point-ring-bottom"
    style="--breakpoint-colour: rbg(125, 67, 100);"
  />
</DocumentFragment>
`;

exports[`Filtered Graph Break Point > should render a [bottom] break point correctly when the theme is [slash] 1`] = `
<DocumentFragment>
  <div
    class="Slash Slash--bottom"
    data-testid="graph-break-point-slash-bottom"
    style="--breakpoint-colour: rbg(125, 67, 100);"
  />
</DocumentFragment>
`;

exports[`Filtered Graph Break Point > should render a [bottom] break point correctly when the theme is [zig-zag] 1`] = `
<DocumentFragment>
  <div
    class="ZigZag ZigZag--bottom"
    data-testid="graph-break-point-zig-zag-bottom"
    style="--breakpoint-colour: rbg(125, 67, 100);"
  />
</DocumentFragment>
`;

exports[`Filtered Graph Break Point > should render a [top] break point correctly when the theme is [dot] 1`] = `
<DocumentFragment>
  <div
    class="Dot Dot--top"
    data-testid="graph-break-point-dot-top"
    style="--breakpoint-colour: rbg(125, 67, 100);"
  />
</DocumentFragment>
`;

exports[`Filtered Graph Break Point > should render a [top] break point correctly when the theme is [double-line] 1`] = `
<DocumentFragment>
  <div
    class="DoubleLine DoubleLine--top"
    data-testid="graph-break-point-double-line-top"
    style="--breakpoint-colour: rbg(125, 67, 100);"
  />
</DocumentFragment>
`;

exports[`Filtered Graph Break Point > should render a [top] break point correctly when the theme is [line] 1`] = `
<DocumentFragment>
  <div
    class="Line Line--top"
    data-testid="graph-break-point-line-top"
    style="--breakpoint-colour: rbg(125, 67, 100);"
  />
</DocumentFragment>
`;

exports[`Filtered Graph Break Point > should render a [top] break point correctly when the theme is [ring] 1`] = `
<DocumentFragment>
  <div
    class="Ring Ring--top"
    data-testid="graph-break-point-ring-top"
    style="--breakpoint-colour: rbg(125, 67, 100);"
  />
</DocumentFragment>
`;

exports[`Filtered Graph Break Point > should render a [top] break point correctly when the theme is [slash] 1`] = `
<DocumentFragment>
  <div
    class="Slash Slash--top"
    data-testid="graph-break-point-slash-top"
    style="--breakpoint-colour: rbg(125, 67, 100);"
  />
</DocumentFragment>
`;

exports[`Filtered Graph Break Point > should render a [top] break point correctly when the theme is [zig-zag] 1`] = `
<DocumentFragment>
  <div
    class="ZigZag ZigZag--top"
    data-testid="graph-break-point-zig-zag-top"
    style="--breakpoint-colour: rbg(125, 67, 100);"
  />
</DocumentFragment>
`;
